conu/apidefs/filesystem.py

```diff
@@ -0,0 +1,141 @@
+import logging
+import os
+import shutil
+
+from tempfile import mkdtemp
+
+import xattr
+
+
+logger = logging.getLogger(__name__)
+
+
+class Filesystem(object):
+    """
+    Utility methods used to access filesystem of containers and images.
+
+    Implementations should probably be done using context managers.
+    """
+    def __init__(self, object_instance, mount_point=None):
+        """
+        :param object_instance: instance of the container or image
+        :param mount_point: str, directory where the filesystem will be mounted
+        """
+        self.obj = object_instance
+        self._mount_point = mount_point
+        self.mount_point_provided = False
+
+    @property
+    def mount_point(self):
+        if self._mount_point is None:
+            self._mount_point = mkdtemp(prefix="conu")
+            self.mount_point_provided = True
+        return self._mount_point
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.mount_point_provided:
+            os.rmdir(self.mount_point)
+
+    def p(self, path):
+        """
+        provide absolute path within the container
+
+        :param path: path with container
+        :return: str
+        """
+        if path.startswith("/"):
+            path = path[1:]
+        p = os.path.join(self.mount_point, path)
+        logger.debug("path = %s", p)
+        return p
+
+    def copy_to(self, src, dest):
+        """
+        copy a file or a directory from host system to a container -- don't implement for images,
+        those are immutable
+
+        :param src: str, path to a file or a directory on host system
+        :param dest: str, path to a file or a directory within container
+        :return: None
+        """
+        raise NotImplementedError("copy_to method is not implemented")
+
+    def copy_from(self, src, dest):
+        """
+        copy a file or a directory from container or image to host system
+
+        :param src: str, path to a file or a directory within container or image
+        :param dest: str, path to a file or a directory on host system
+        :return: None
+        """
+        p = self.p(src)
+        if os.path.isfile(src):
+            logger.info("copying file %s to %s", p, dest)
+            shutil.copy2(p, dest)
+        else:
+            logger.info("copying directory %s to %s", p, dest)
+            shutil.copytree(p, dest)
```

Comment:
What about if directory already exists on the filesystem? Please check whether destination does not exist or return as an exception.

Reply:
I agree that I didn't take care of error handling much.

Please provide an example, as I don't understand what use case are we talking about here. Ideally in a form of a test.

Reply:
Like is mentioned in docu https://docs.python.org/2/library/shutil.html. Especially `The destination directory, named by dst, must not already exist; it will be created as well as missing parent directories. `

Let's say user wants to copy directory `/var/tmp/abc` to container `/tmp/abc`. But it fails if `/tmp/abc` already exists. Nothing more.

Reply:
...and it fails with:
 * `OSError` on python 2: `OSError: [Errno 17] File exists: '/tmp/pytest-of-root/pytest-1/test_copy_from0'`
 * `FileExistsError` on python 3: `FileExistsError: [Errno 17] File exists: '/tmp/pytest-of-root/pytest-0/test_copy_from0'`

I still don't understand what the issue is. Can you please share with me how you propose to solve it?

Reply:
I improved the docs and added more tests. PTAL

Reply:
Docs are fine.

I thought something like that.
~~~
14:35 $ ls ~/temp2/
tests
✔ ~/work/container-images/memcached [master|✚ 2…3] 
14:35 $ ipython
Python 2.7.13 (default, Sep  5 2017, 08:53:59) 
Type "copyright", "credits" or "license" for more information.

IPython 5.3.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import os

In [2]: import shutil

In [3]: shutil.copytree("/home/phracek/android-studio/", "/home/phracek/temp2/")
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-3-eaa53d5580ad> in <module>()
----> 1 shutil.copytree("/home/phracek/android-studio/", "/home/phracek/temp2/")

/usr/lib64/python2.7/shutil.pyc in copytree(src, dst, symlinks, ignore)
    175         ignored_names = set()
    176 
--> 177     os.makedirs(dst)
    178     errors = []
    179     for name in names:

/usr/lib64/python2.7/os.pyc in makedirs(name, mode)
    155         if tail == curdir:           # xxx/newdir/. exists if xxx/newdir exists
    156             return
--> 157     mkdir(name, mode)
    158 
    159 def removedirs(name):

OSError: [Errno 17] File exists: '/home/phracek/temp2/'


~~~

And before shutil.copytree I would check dest directory if it does not exist.
And if exists, then return either conuexception or False.

Reply:
Question is if we want to return True and False or tester should care about OSError exceptions.
That was my point.

Reply:
I would not return anything: if the operation failed, raise an exception (with proper error message), if it went fine, just move on. That's why I think that the current behaviour is okay.